### PR TITLE
Improve blackjax sampling integration

### LINKED ISSUE #6550: BUG: sample_blackjax_nuts incorrectly reports sampling time
### Describe the issue:

For larger models, `sample_blackjax_nuts` reports (unfortunately) unrealistically small sampling time and a very large transformation time. This is due to JAX's [Asynchronous dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html#async-dispatch) model. I'd guess what's currently reported as "sampling time" is really JAX compilation. It should be able to be fixed with a simple `block_until_ready()`. I can send a PR shortly.

Mostly a cosmetic issue, but it caused me some time trying to work out why transformation was so slow 😄.

### Reproduceable code example:

```python
import pymc as pm
from pymc.sampling.jax import sample_blackjax_nuts
import numpy as np

x = np.random.normal(0, 1, 1_000_000)
with pm.Model() as model:
    sigma = pm.HalfCauchy("sigma", 1)
    mu = pm.Normal("mu", 0, 1)
    pm.Normal("x", mu, sigma, observed=x)
    trace_blackjax = sample_blackjax_nuts()
```

### Error message:

```shell
Compiling...
Compilation time =  0:00:02.020721
Sampling...
Sampling time =  0:00:09.755916
Transforming variables...
Transformation time =  0:02:29.152843
```

### PyMC version information:

macOS

```
pytensor            2.9.1
blackjax            0.9.6
pymc                5.0.2+58.g33d641db
jax                 0.4.1
jaxlib              0.4.1

```

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by junpenglao**
Closes #6550 

**What is this PR about?**
Adding more fine tune control in blackjax sampling.

After this PR:
```python
# Use full rank mass matrix in window adaptation 
with model:
    idata2 = pm.sample(
        nuts_sampler="blackjax", 
        nuts_sampler_kwargs={
            "chain_method": "vectorized",
            "adaptation_kwargs": {
                "is_mass_matrix_diagonal": False
                }
            }
        )

# Warm up and sample with HMC (instead of NUTS) with full rank mass matrix adaptation
with model:
    idata2 = pm.sample(
        nuts_sampler="blackjax", 
        nuts_sampler_kwargs={
            "chain_method": "vectorized",
            "adaptation_kwargs": {
                "is_mass_matrix_diagonal": False,
                "algorithm": "hmc",
                "num_integration_steps": 20,
                }
            }
        )

# Enable progress bar (only works for `chain_method =  "vectorized"`)
with model:
    idata2 = pm.sample(
        nuts_sampler="blackjax", 
        progress_bar=True,
        nuts_sampler_kwargs={
            "chain_method": "vectorized",
            # "chain_method": "parallel",
            "adaptation_kwargs": {
                "is_mass_matrix_diagonal": False,
                }
            }
        )
```

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [ ] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
None

## New features
More option in Blackjax sampling

## Bugfixes
None

## Documentation
None

## Maintenance
None

----
:books: Documentation preview :books:: https://pymc--6963.org.readthedocs.build/en/6963/

### CODE REVIEW

File: pymc/sampling/jax.py [hunk 1]
```diff
@@ -289,25 +289,22 @@ def _update_coords_and_dims(
...
-    algorithm=None,
+    seed, init_position, logprob_fn, draws, tune, target_accept, **adaptation_kwargs
 ):
     import blackjax
 
-    if algorithm is None:
+    algorithm_name = adaptation_kwargs.pop("algorithm", "nuts")
+    if algorithm_name == "nuts":
         algorithm = blackjax.nuts
+    else:
```
Comment (ricardoV94) [inline pymc/sampling/jax.py]:
```suggestion
    elif algorithm_name == "hmc":
```

And raise in the else?

Comment (junpenglao) [reply pymc/sampling/jax.py]: Done, thanks

**Comment by junpenglao** [comment]
@jessegrabowski FYI the progress bar is partially working for both warm up and sampling.

**Review by twiecki — APPROVED** [review]

File: pymc/sampling/jax.py [hunk 2]
```diff
@@ -432,36 +441,48 @@ def sample_blackjax_nuts(
...
-
-    print("Sampling...", file=sys.stdout)
+    if adaptation_kwargs is None:
+        adaptation_kwargs = {}
 
     # Adapted from numpyro
     if chain_method == "parallel":
         map_fn = jax.pmap
+        if adaptation_kwargs.get("progress_bar", False):
+            import warnings
```
Comment (ricardoV94) [inline pymc/sampling/jax.py:451]: This should respect the `progress_bar` from `pm.sample`. Does it?

File: pymc/sampling/jax.py [hunk 3]
```diff
@@ -289,40 +291,46 @@ def _update_coords_and_dims(
...
-            state, info = kernel(rng_key, state)
-            return state, (state, info)
+    def _one_step(state, xs):
+        _, rng_key = xs
+        state, info = kernel(rng_key, state)
+        return state, (state, info)
 
-        keys = jax.random.split(rng_key, draws)
-        _, (states, infos) = jax.lax.scan(one_step, initial_state, keys)
+    progress_bar = adaptation_kwargs.pop("progress_bar", False)
```
Comment (ricardoV94) [inline pymc/sampling/jax.py:321]: This should follow the `progressbar` whet called via `pm.sample`, but I suspect it now comes from a separate argument?

Comment (ricardoV94) [reply pymc/sampling/jax.py:321]: Specifically here we could put `progress_bar` in the right place: https://github.com/pymc-devs/pymc/blob/009da35cb608f254cfc82b86513054d1c56ab992/pymc/sampling/mcmc.py#L367

You'll see that the other two methods propagate `progressbar`. Also watch out for the underscore. PyMC doesn't have one, but blackjax seems to.

Comment (junpenglao) [reply pymc/sampling/jax.py:321]: Make sense, thanks!

Also:
```
pymc: progressbar
nutpie: progress_bar
numpyro: progressbar
blackjax: progress_bar
```

**Review by ricardoV94 — APPROVED** [review]

**Pull request merged**